datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -142,6 +151,7 @@ public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
           // Make sure to replace the DatastreamTask with most up to date info
           // This is necessary because DatastreamTaskImpl.hashCode() does not take into account all the
           // fields/properties of the DatastreamTask (e.g. dependencies).
+          _runningTasks.remove(task);
```

Comment:
This is not a problem. We are not iterating through the map that we are removing things from. We are iterating over the tasks list, which we don't modify at all.

      **for (DatastreamTask task : tasks) {**
        ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(task);
        if (connectorTaskEntry != null) {
          AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = connectorTaskEntry.getConnectorTask();
          kafkaBasedConnectorTask.checkForUpdateTask(task);
          // Make sure to replace the DatastreamTask with most up to date info
          // This is necessary because DatastreamTaskImpl.hashCode() does not take into account all the
          // fields/properties of the DatastreamTask (e.g. dependencies).
          **_runningTasks.remove(task);**
          _runningTasks.put(task, connectorTaskEntry);
          continue; // already running
        }

        _runningTasks.put(task, createKafkaConnectorTask(task));
      }